### ISSUE TITLE
assumptions

### ISSUE THREAD

**Issue by thejohnfreeman** [sampsyo/confit#7]
### Background

I wrote a configuration library in Python for my employer before discovering Confit. They share many of the same features, but I think there are some things mine does better than Confit. Instead of spawning a new project, I'd like to contribute to Confit, but not without finding a consensus on philosophy, features, and implementation. I'm submitting a few different proposals as separate Github issues.

### Assumptions

I want to start by discussing a few assumptions I used.

1. **A configuration consists of variable bindings.** In each new project, a configuration is born only when some variable holds a magic value (e.g. hostname of external service, port to bind, path to temporary directory, path to credentials). The magic value is moved to the configuration as a default, and the variable's value is loaded from the configuration. There is no need for structure richer than a mapping of names to values.

2. **Each variable has a type.** Equivalently, each variable can be validated.

3. **Each type can be serialized to and deserialized from a string.** Configuration files, environment variables, and command line arguments are just strings. We have to be able to represent values as strings.

Can we adopt these assumptions? Do we have real-world counterexamples that cannot be conveniently expressed with these assumptions in place? I'd like to hear your thoughts.

**Comment by hkhanna** [comment]
John we played kickball together! I knew I recognized you. 

Although i'll let sampyso speak to the substance of your issues, I think you'll enjoy hacking on his projects. He's a great person to work with.

**Comment by thejohnfreeman** [comment]
haha nice :)

**Comment by sampsyo** [comment]
These all seem reasonable from a high level. I'd of course be interested in talking further about the details—here are some thoughts that may or may not match with what you were expecting. (I'll refer to [beets](https://github.com/sampsyo/beets) several times, which is the main way I use Confit now.)

On structure: Yes, replacing variables seems like a good philosophy for configuration. That said, there are some ways that certain kinds of structure are useful. One simple way: In a very large application, it can make sense for UI reasons to group thematically related options together and avoid namespace collisions. For example, this config file in YAML syntax:

    musicbrainz:
        user: foo
        password: bar
    discogs:
        user: baz
        password: qux

can be more legible that a completely flat key/value mapping. I'd love to hear whether this feels like too much structure to you.

On typing and serialization: Yes, this seems right. It's worthwhile to remember that validation and deserialization are often the same process (or at least interrelated). Another priority for me in evolving Confit is to better support user-defined validation and conversion.

**Comment by thejohnfreeman** [comment]
That's not too much structure. I'd characterize it as a variable binding whose value is an object/mapping/record. The key distinction is that I wouldn't care to support overriding just the username or password alone. Users would always have to override both at once, but in the case of username+password pairs, that seems reasonable. Would you agree? Is there another counterexample?

**Comment by sampsyo** [comment]
Ah, good question. I don't have a counterexample for your policy at the moment, so let's go with that for now.

FWIW, the current Confit does support overriding at arbitrary points in a hierarchy of dictionary and lists. This works just by trying to look up a given field in each (potentially sparse) structure. So a source that contains `{'musicbrainz': {'password': 'foo'}}` effectively overrides the password and not the user.

**Comment by thejohnfreeman** [comment]
Yeah, I thought that was a really cool feature when I first found Confit, but I'm not sure how much people will need it, and I don't think it plays well with overrides from the environment or command line, or with validation. I'm leaning toward considering it a complication of the model and YAGNI, but I can be convinced otherwise.

**Comment by sampsyo** [comment]
Great. Let's keep an open mind, but for the moment, I'm willing to believe we're better off arbitrary without deep overlays like that.

**Comment by thejohnfreeman** [comment]
I think I'm ready to get started with some writing. How about I start with a README for a hypothetical version of the library that outlines the concepts with example uses to demonstrate all of the features? I'll have an implementation in mind, but we can poke at the API first until we agree on something. Should I add it as a file in this repo? Edit the docs in a branch?

**Comment by sampsyo** [comment]
Yes, sounds great! I'm adding you as a collaborator. Let's start work in a new branch. That way, we can prototype tests and pieces of the implementation before the API is completely settled.

**Comment by sampsyo** [comment]
OK, here's one slightly more concrete example of fine-grained overriding. If instead you have a section for configuring the "server" component of your application:

    server:
        host: 0.0.0.0
        port: 8337

then you probably want to override these values independently using command-line flags, as in:

    $ some_command serve --port 8338

but this would be tricky if you had to override the whole top-level `server` variable using the results from your command-line parser.

In general, the idea is that you would want to treat `host` under `server` (or `user` under `musicbrainz`) as a first-class variable, with the same rights and privileges to type-validation and overriding as others. I can't think of any situations where this arises other than when defining *sections* in configuration files, but this can actually be a useful pattern—even if they're not infinitely nestable. (My `~/.gitconfig` and `~/.hgrc` are full of sections, for example, for usability and namespacing purposes.)

This is another question I don't have an answer to, but: How can we cleanly support sectioned configurations? The current Confit does it by letting the data be completely free-form and allowing validation and overriding at any point in the hierarchy. The new proposal goes to the opposite extreme, allowing overriding and validation only for top-level keys. Maybe there's something in the middle that is both simple to specify but flexible enough to allow basic hierarchy?

**Comment by thejohnfreeman** [comment]
That's where I imagined we'd use scoping string-encoded in the variable names:

```
server.host = example.com
server.port = 80
```

In my first draft, you'll see mentions of separators in the `from_env` and `from_ini` sources. I intended those to be used so that, for example, `APP_SERVER_HOST` in the environment would override the variable `server.host` if the app called `from_env(prefix='APP_', sep='.')`. I forgot to add it in my first draft, but I could see this pattern working for command line args as well, e.g. `--app-server-host` overrides `server.host` if `from_args(prefix='app-', sep='.')` is called.

**Comment by sampsyo** [comment]
Aha, sounds interesting! Sounds like a good approach. I'm also interested to play around with this kind of aliasing to see how it works with modularity too—for example, can it be easy for the "server" component to somehow flatten this to `server_config.host` as opposed to `global_config.server.host` in the code?

I'm not quite sure how this would work in YAML, where dictionaries could either indicate sections or actual key/value maps as values.

**Comment by thejohnfreeman** [comment]
I think we can certainly add that type of projection, should be easy. I'll think more about nesting regardless.

**Comment by jrobeson** [comment]
I dev on another project that uses env variables to override configuration. it ends up looking like this:

`FOO__BAR__BAZ_QUUX`

```yaml
foo:
  bar:
    baz_quux:
```

thus `__` is used for nesting
